### PR TITLE
Use set-based interface for NodesStatsRequest

### PR DESCRIPTION
**Pull request by williamrandolph**
This PR refactors the NodesStatsRequest object towards pluggability. It has the same basic structure as a similar refactor of the NodesInfoRequest object.

Now that the `NodesStatsRequest` class uses a set of strings for its internal serialization, we need to update the class's interface so that we no longer use hard-coded getters and setters, but rather methods that add strings directly. For example, the old way of adding "os" metrics to a request would be to call `request.os(true)`. The new way of doing this is to call `request.addMetric("os")`.

For the time being, the canonical list of metrics is an enum in `NodesStatsRequest`. This will eventually be replaced with something pluggable.

Note that the `NodesStatsRequestBuilder` class also needs to be refactored, but I'm leaving that for a later PR in order to keep the diff manageable.

Relates #52975 (meta issue)
Relates #53410 (parallel refactor of Nodes Info)
Relates #53235 (previous work on NodesStatsRequest)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java [hunk 1]
```diff
@@ -114,178 +130,65 @@ public NodesStatsRequest indices(boolean indices) {
...
+     * Add an array of metric names
      */
-    public NodesStatsRequest ingest(boolean ingest) {
-        addOrRemoveMetric(ingest, Metric.INGEST.metricName());
+    public NodesStatsRequest addMetrics(String... metrics) {
+        // use sorted set for reliable ordering in error messages
+        SortedSet<String> illegalMetrics = new TreeSet<>();
+        Set<String> validMetrics = new HashSet<>();
+        for (String metric : metrics) {
+            if (Metric.allMetrics().contains(metric)) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java]: Could we simplify this by using `Set.containsAll`? Essentially this:
```
Set<String> metricsSet = new TreeSet<>(metrics);
if (Metric.allMetrics().containsAll(metricSet) == false) {
    Set<String> illegalMetrics = metricsSet.removaAll(Metric.allMetrics());
    // throw
}
requestMetrics.addAll(metricsSet);
```

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java [hunk 2]
```diff
@@ -311,11 +214,20 @@ public void writeTo(StreamOutput out) throws IOException {
         }
     }
 
+    public static Map<String, Consumer<NodesStatsRequest>> getDispatchMap() {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java]: Why is this moved to here? It seems like an implementation detail that should stay in the stats action. The request should be a opaque list of strings?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java [hunk 3]
```diff
@@ -327,15 +239,15 @@ public void writeTo(StreamOutput out) throws IOException {
         SCRIPT("script"),
         DISCOVERY("discovery"),
         INGEST("ingest"),
-        ADAPTIVE_SELECTION("adaptiveSelection");
+        ADAPTIVE_SELECTION("adaptiveSelection"); // TODO: Should this be removed?
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java]: Can you elaborate on why this might need to be removed?

**Review by rjernst — COMMENTED** [review]
A couple minor questions/suggestions

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java (hunk 1 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java]: Ah, that looks better. I'll do this.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java (hunk 3 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java]: It definitely doesn't need to be removed. I'll remove the `TODO`, which was a reminder for me to look into it.

Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequest.java]: Oh, darn. I do need to change it to `adaptive_selection`, though.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by williamrandolph** [comment]
Backport to 7.x: https://github.com/elastic/elasticsearch/pull/54223
Small bit of this backported to 7.7: https://github.com/elastic/elasticsearch/pull/54193
